### PR TITLE
Feat [CHK-3511] Adding checkpointing service logic and configuration

### PR DESCRIPTION
**Pull request by EmanueleBVtech**
#### List of Changes
- Added checkpointing service RedisResumePolicyService
- Added configuration update in HELM files
- Updated application properties
- Updated README
- Added unit tests

#### Motivation and Context
We need to add a checkpointing feature to make sure we can restart processing events in case of disservice of any kind.

#### How Has This Been Tested?
Unit tests

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
  expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: helm/pay-wallet-values-uat.yaml [hunk 1]
```diff
@@ -103,6 +103,11 @@ microservice-chart:
     SCHEDULER_CDC_QUEUE_VISIBILITY_TIMEOUT_SECONDS: "0"
     SCHEDULER_CDC_SEND_RETRY_MAX_ATTEMPTS: "1"
     SCHEDULER_CDC_SEND_RETRY_INTERVAL_IN_MS: "1000"
+    CDC_REDIS_RESUME_KEYSPACE: "pay-wallet-scheduler"
+    CDC_REDIS_RESUME_TARGET: "pay-wallet-job"
+    CDC_REDIS_RESUME_FALLBACK_IN_MIN: "60"
+    CDC_REDIS_RESUME_TTL_IN_MIN: "120"
+    CDC_RESUME_SAVE_INTERVAL: "5"
```
Comment (CianoDanilo) [inline helm/pay-wallet-values-uat.yaml:110]: Please remove save interval, the job will be responsible to handle the configuration for saving checkpoints.
Please consider using the prefix SCHEDULER_ and removing CDC_REDIS_RESUME_TARGET, the job id will be passed to service's method as an input parameter

File: src/main/resources/application.properties [hunk 2]
```diff
@@ -23,4 +23,10 @@ scheduler.cdc.queue.ttlSeconds=\${SCHEDULER_CDC_QUEUE_TTL_SECONDS}
...
 scheduler.cdc.retry-send.maxAttempts=\${SCHEDULER_CDC_SEND_RETRY_MAX_ATTEMPTS}
-scheduler.cdc.retry-send.intervalInMs=\${SCHEDULER_CDC_SEND_RETRY_INTERVAL_IN_MS}
\ No newline at end of file
+scheduler.cdc.retry-send.intervalInMs=\${SCHEDULER_CDC_SEND_RETRY_INTERVAL_IN_MS}
+# CDC resume policy
+cdc.redis-resume.keyspace=\${CDC_REDIS_RESUME_KEYSPACE}
+cdc.redis-resume.target=\${CDC_REDIS_RESUME_TARGET}
+cdc.redis-resume.fallbackInMin=\${CDC_REDIS_RESUME_FALLBACK_IN_MIN}
+cdc.redis-resume.ttlInMin=\${CDC_REDIS_RESUME_TTL_IN_MIN}
+cdc.resume.saveInterval=\${CDC_RESUME_SAVE_INTERVAL}
```
Comment (CianoDanilo) [inline src/main/resources/application.properties:32]: Please remove save interval, the job will be responsible to handle the configuration for saving checkpoints.
Please consider using the prefix scheduler. for these properties and removing cdc.redis-resume.target, the job id will be passed to service's method as an input parameter

File: src/main/kotlin/it/pagopa/wallet/scheduler/service/RedisResumePolicyService.kt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import org.springframework.stereotype.Service
+
+@Service
+class RedisResumePolicyService(
+    @Autowired private val redisTemplate: TimestampRedisTemplate,
+    @Autowired private val redisResumePolicyConfig: RedisResumePolicyConfig
+) : ResumePolicyService {
+    private val logger = LoggerFactory.getLogger(RedisResumePolicyService::class.java)
+
+    override fun getResumeTimestamp(): Instant {
```
Comment (CianoDanilo) [inline src/main/kotlin/it/pagopa/wallet/scheduler/service/RedisResumePolicyService.kt:18]: Consider adding a String as method's input to pass to findByKeyspaceAndTarget method as second parameter

File: src/main/kotlin/it/pagopa/wallet/scheduler/service/RedisResumePolicyService.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            .orElseGet {
+                logger.warn(
+                    "Resume timestamp not found on Redis, fallback on Instant.now()-{} minutes",
+                    redisResumePolicyConfig.fallbackInMin
+                )
+                Instant.now().minus(redisResumePolicyConfig.fallbackInMin, ChronoUnit.MINUTES)
+            }
+    }
+
+    override fun saveResumeTimestamp(timestamp: Instant) {
```
Comment (CianoDanilo) [inline src/main/kotlin/it/pagopa/wallet/scheduler/service/RedisResumePolicyService.kt:33]: Consider adding a String as method's input to pass to save method as second parameter

File: helm/pay-wallet-values-dev.yaml [hunk 5]
```diff
@@ -81,6 +81,11 @@ microservice-chart:
     SCHEDULER_CDC_QUEUE_VISIBILITY_TIMEOUT_SECONDS: "0"
     SCHEDULER_CDC_SEND_RETRY_MAX_ATTEMPTS: "1"
     SCHEDULER_CDC_SEND_RETRY_INTERVAL_IN_MS: "1000"
+    CDC_REDIS_RESUME_KEYSPACE: "pay-wallet-scheduler"
+    CDC_REDIS_RESUME_TARGET: "pay-wallet-job"
+    CDC_REDIS_RESUME_FALLBACK_IN_MIN: "10"
+    CDC_REDIS_RESUME_TTL_IN_MIN: "120"
+    CDC_RESUME_SAVE_INTERVAL: "5"
```
Comment (CianoDanilo) [inline helm/pay-wallet-values-dev.yaml:88]: Please remove save interval, the job will be responsible to handle the configuration for saving checkpoints.
Please consider using the prefix SCHEDULER_ for these variables and removing CDC_REDIS_RESUME_TARGET, the job id will be passed to service's method as an input parameter

File: src/main/kotlin/it/pagopa/wallet/scheduler/service/RedisResumePolicyService.kt (hunk 4 above)
Comment (EmanueleBVtech) [reply src/main/kotlin/it/pagopa/wallet/scheduler/service/RedisResumePolicyService.kt]: done it here 13e819008aa9a134136c7ae5ba42caba884e9c93

File: src/main/kotlin/it/pagopa/wallet/scheduler/service/RedisResumePolicyService.kt (hunk 3 above)
Comment (EmanueleBVtech) [reply src/main/kotlin/it/pagopa/wallet/scheduler/service/RedisResumePolicyService.kt]: done it here 13e819008aa9a134136c7ae5ba42caba884e9c93

File: helm/pay-wallet-values-dev.yaml (hunk 5 above)
Comment (EmanueleBVtech) [reply helm/pay-wallet-values-dev.yaml]: done it here 13e819008aa9a134136c7ae5ba42caba884e9c93

File: src/main/resources/application.properties (hunk 2 above)
Comment (EmanueleBVtech) [reply src/main/resources/application.properties]: done it here 1acaceec11c78fc66778f1ef4f816f0dc0e1a1a7

File: src/main/kotlin/it/pagopa/wallet/scheduler/service/RedisResumePolicyService.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+            .findByKeyspaceAndTarget(
+                redisResumePolicyConfig.keyspace,
+                redisResumePolicyConfig.target
+            )
+            .orElseGet {
+                logger.warn(
+                    "Resume timestamp not found on Redis, fallback on Instant.now()-{} minutes",
+                    redisResumePolicyConfig.fallbackInMin
+                )
+                Instant.now().minus(redisResumePolicyConfig.fallbackInMin, ChronoUnit.MINUTES)
```
[the remark this answers is not in the document]
Comment (EmanueleBVtech) [reply src/main/kotlin/it/pagopa/wallet/scheduler/service/RedisResumePolicyService.kt]: done

File: src/main/kotlin/it/pagopa/wallet/scheduler/service/RedisResumePolicyService.kt [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+class RedisResumePolicyService(
+    @Autowired private val redisTemplate: TimestampRedisTemplate,
+    @Autowired private val redisResumePolicyConfig: RedisResumePolicyConfig
+) : ResumePolicyService {
+    private val logger = LoggerFactory.getLogger(RedisResumePolicyService::class.java)
+
+    override fun getResumeTimestamp(target: String): Instant {
+        return redisTemplate
+            .findByKeyspaceAndTarget(redisResumePolicyConfig.keyspace, target)
+            .orElseThrow { NoSuchElementException("No resume timestamp found for target: $target") }
```
Comment (CianoDanilo) [inline src/main/kotlin/it/pagopa/wallet/scheduler/service/RedisResumePolicyService.kt:21]: Do not throw an exception if no checkpoint is found, it is acceptable not having it (e.g. first execution of the job); we could consider changing getResumeTimestamp return type to match the return type of findByKeyspaceAndTarget method (Optional<Instant>).

File: src/main/resources/application.properties [hunk 8]
```diff
@@ -23,4 +23,8 @@ scheduler.cdc.queue.ttlSeconds=\${SCHEDULER_CDC_QUEUE_TTL_SECONDS}
 scheduler.cdc.queue.visibilityTimeoutWalletCdc=\${SCHEDULER_CDC_QUEUE_VISIBILITY_TIMEOUT_SECONDS}
 # Retry policy for sending message
 scheduler.cdc.retry-send.maxAttempts=\${SCHEDULER_CDC_SEND_RETRY_MAX_ATTEMPTS}
-scheduler.cdc.retry-send.intervalInMs=\${SCHEDULER_CDC_SEND_RETRY_INTERVAL_IN_MS}
\ No newline at end of file
+scheduler.cdc.retry-send.intervalInMs=\${SCHEDULER_CDC_SEND_RETRY_INTERVAL_IN_MS}
+# CDC resume policy
+cdc.redis-resume.keyspace=\${SCHEDULER_REDIS_RESUME_KEYSPACE}
+cdc.redis-resume.fallbackInMin=\${SCHEDULER_REDIS_RESUME_FALLBACK_IN_MIN}
+cdc.redis-resume.ttlInMin=\${SCHEDULER_REDIS_RESUME_TTL_IN_MIN}
```
Comment (CianoDanilo) [inline src/main/resources/application.properties:30]: Change the properties name according to reflect what has been done for helm env var (cdc -> scheduler)

File: helm/pay-wallet-values-dev.yaml [hunk 9]
```diff
@@ -81,6 +81,9 @@ microservice-chart:
     SCHEDULER_CDC_QUEUE_VISIBILITY_TIMEOUT_SECONDS: "0"
     SCHEDULER_CDC_SEND_RETRY_MAX_ATTEMPTS: "1"
     SCHEDULER_CDC_SEND_RETRY_INTERVAL_IN_MS: "1000"
+    SCHEDULER_REDIS_RESUME_KEYSPACE: "pay-wallet-scheduler"
+    SCHEDULER_REDIS_RESUME_FALLBACK_IN_MIN: "10"
```
Comment (CianoDanilo) [inline helm/pay-wallet-values-dev.yaml:85]: SCHEDULER_REDIS_RESUME_FALLBACK_IN_MIN is not needed.

File: helm/pay-wallet-values-uat.yaml [hunk 10]
```diff
@@ -103,6 +103,9 @@ microservice-chart:
     SCHEDULER_CDC_QUEUE_VISIBILITY_TIMEOUT_SECONDS: "0"
     SCHEDULER_CDC_SEND_RETRY_MAX_ATTEMPTS: "1"
     SCHEDULER_CDC_SEND_RETRY_INTERVAL_IN_MS: "1000"
+    SCHEDULER_REDIS_RESUME_KEYSPACE: "pay-wallet-scheduler"
+    SCHEDULER_REDIS_RESUME_FALLBACK_IN_MIN: "60"
```
Comment (CianoDanilo) [inline helm/pay-wallet-values-uat.yaml:107]: SCHEDULER_REDIS_RESUME_FALLBACK_IN_MIN is not needed.

File: src/main/resources/application.properties [hunk 11]
```diff
@@ -23,4 +23,8 @@ scheduler.cdc.queue.ttlSeconds=\${SCHEDULER_CDC_QUEUE_TTL_SECONDS}
 scheduler.cdc.queue.visibilityTimeoutWalletCdc=\${SCHEDULER_CDC_QUEUE_VISIBILITY_TIMEOUT_SECONDS}
 # Retry policy for sending message
 scheduler.cdc.retry-send.maxAttempts=\${SCHEDULER_CDC_SEND_RETRY_MAX_ATTEMPTS}
-scheduler.cdc.retry-send.intervalInMs=\${SCHEDULER_CDC_SEND_RETRY_INTERVAL_IN_MS}
\ No newline at end of file
+scheduler.cdc.retry-send.intervalInMs=\${SCHEDULER_CDC_SEND_RETRY_INTERVAL_IN_MS}
+# CDC resume policy
+cdc.redis-resume.keyspace=\${SCHEDULER_REDIS_RESUME_KEYSPACE}
+cdc.redis-resume.fallbackInMin=\${SCHEDULER_REDIS_RESUME_FALLBACK_IN_MIN}
```
Comment (CianoDanilo) [inline src/main/resources/application.properties:29]: cdc.redis-resume.fallbackInMin is not needed.

File: helm/pay-wallet-values-dev.yaml (hunk 9 above)
Comment (EmanueleBVtech) [reply helm/pay-wallet-values-dev.yaml]: done here a963bf41949734565b131774a0da0809fc4bfc61

File: helm/pay-wallet-values-uat.yaml (hunk 10 above)
Comment (EmanueleBVtech) [reply helm/pay-wallet-values-uat.yaml]: done here a963bf41949734565b131774a0da0809fc4bfc61

File: src/main/resources/application.properties (hunk 11 above)
Comment (EmanueleBVtech) [reply src/main/resources/application.properties]: done here a963bf41949734565b131774a0da0809fc4bfc61

File: src/main/resources/application.properties (hunk 8 above)
Comment (EmanueleBVtech) [reply src/main/resources/application.properties]: done here a963bf41949734565b131774a0da0809fc4bfc61

File: src/main/kotlin/it/pagopa/wallet/scheduler/service/RedisResumePolicyService.kt (hunk 7 above)
Comment (EmanueleBVtech) [reply src/main/kotlin/it/pagopa/wallet/scheduler/service/RedisResumePolicyService.kt:21]: mmm ok, i'm on it

Comment (EmanueleBVtech) [reply src/main/kotlin/it/pagopa/wallet/scheduler/service/RedisResumePolicyService.kt]: done here a12e3b262d3f88eb6af34fb56fc1f48869425118

File: helm/pay-wallet-values-dev.yaml [hunk 12]
```diff
@@ -81,6 +81,8 @@ microservice-chart:
     SCHEDULER_CDC_QUEUE_VISIBILITY_TIMEOUT_SECONDS: "0"
     SCHEDULER_CDC_SEND_RETRY_MAX_ATTEMPTS: "1"
     SCHEDULER_CDC_SEND_RETRY_INTERVAL_IN_MS: "1000"
+    SCHEDULER_REDIS_RESUME_KEYSPACE: "pay-wallet-scheduler"
+    SCHEDULER_REDIS_RESUME_TTL_IN_MIN: "120"
```
Comment (CianoDanilo) [inline helm/pay-wallet-values-dev.yaml:85]:
```suggestion
    SCHEDULER_REDIS_RESUME_TTL_IN_MIN: "120"
    REDIS_HOST: pagopa-d-itn-pay-wallet-redis.redis.cache.windows.net
    REDIS_PORT: "6380"
    REDIS_SSL_ENABLED: "true"
```

File: helm/pay-wallet-values-dev.yaml [hunk 13]
```diff
@@ -81,6 +81,8 @@ microservice-chart:
...
   envSecret:
     OTEL_EXPORTER_OTLP_HEADERS: elastic-otel-token-header
```
Comment (CianoDanilo) [inline helm/pay-wallet-values-dev.yaml:87]:
```suggestion
    OTEL_EXPORTER_OTLP_HEADERS: elastic-otel-token-header
    REDIS_PASSWORD: redis-wallet-password
```

File: helm/pay-wallet-values-uat.yaml [hunk 14]
```diff
@@ -103,6 +103,8 @@ microservice-chart:
     SCHEDULER_CDC_QUEUE_VISIBILITY_TIMEOUT_SECONDS: "0"
     SCHEDULER_CDC_SEND_RETRY_MAX_ATTEMPTS: "1"
     SCHEDULER_CDC_SEND_RETRY_INTERVAL_IN_MS: "1000"
+    SCHEDULER_REDIS_RESUME_KEYSPACE: "pay-wallet-scheduler"
+    SCHEDULER_REDIS_RESUME_TTL_IN_MIN: "120"
```
Comment (CianoDanilo) [inline helm/pay-wallet-values-uat.yaml:107]:
```suggestion
    SCHEDULER_REDIS_RESUME_TTL_IN_MIN: "120"
    REDIS_HOST: pagopa-u-itn-pay-wallet-redis.redis.cache.windows.net
    REDIS_PORT: "6380"
    REDIS_SSL_ENABLED: "true"
```

File: helm/pay-wallet-values-uat.yaml [hunk 15]
```diff
@@ -103,6 +103,8 @@ microservice-chart:
...
   envSecret:
     OTEL_EXPORTER_OTLP_HEADERS: elastic-otel-token-header
```
Comment (CianoDanilo) [inline helm/pay-wallet-values-uat.yaml:109]:
```suggestion
    OTEL_EXPORTER_OTLP_HEADERS: elastic-otel-token-header
    REDIS_PASSWORD: redis-wallet-password
```

File: README.md [hunk 16]
```diff
@@ -61,6 +61,10 @@ If you want to customize the application environment, reference this table:
 | SCHEDULER_CDC_QUEUE_VISIBILITY_TIMEOUT_SECONDS | Visibility timeout in seconds for expired event                                                                                                            |                   |         |
 | SCHEDULER_CDC_SEND_RETRY_MAX_ATTEMPTS          | Max configurable attempts for performing the logic business related to a change event                                                                      | long              |         |
 | SCHEDULER_CDC_SEND_RETRY_INTERVAL_IN_MS        | Configurable interval in milliseconds between retries attempts                                                                                             | long              |         |
```
Comment (CianoDanilo) [inline README.md:63]: Let's add newly added Redis env var

Comment (EmanueleBVtech) [reply README.md]: done here f6457e2d1bb0b07505a7966a1e5504932f829eda

**Review by CianoDanilo — APPROVED** [review]

**Review by pietro-tota — APPROVED** [review]

**Pull request merged**